src/hb-ot-layout-gsubgpos.hh

```diff
@@ -2002,6 +2002,77 @@ struct ChainRule
 					      lookup.arrayZ, lookup_context));
   }
 
+  void serialize (hb_serialize_context_t *c,
+                  const hb_map_t *backtrack_map,
+                  const hb_map_t *input_map = nullptr,
+                  const hb_map_t *lookahead_map = nullptr) const
+  {
+    const hb_map_t *mapping = backtrack_map;
+    if (unlikely (!c->allocate_size<ArrayOf<HBUINT16>> (backtrack.get_size ()))) return;
+    c->copy (backtrack.len);
+    for (const unsigned g : + backtrack.iter ()
+			    | hb_map (mapping))
+    {
+      HBUINT16 gid;
+      gid = g;
+      c->copy (gid);
+    }
+
+    if (unlikely (!c->allocate_size<HeadlessArrayOf<HBUINT16>> (inputX.get_size ()))) return;
+    c->copy (inputX.lenP1);
+    if (input_map) mapping = input_map;
```

Comment:
Is it intentional that backtrack mapping is used if input_map is not specified?

Reply:
yes.
format2 has different mappings for backtrack, input and lookahead arrays: 
       old class -> new class 
format1, mappings are the same for all arrays: 
        old glyph id -> new glyph id,

Reply:
sounds good.